### PR TITLE
fix(qc-research): repair frozen PPO actor in research_rl_ppo (#3326)

### LINKED ISSUE #3326: [#3164] research_rl_ppo: PPO actor fige a l'init (bug total_loss) — verdict NO BEATS mesure une policy aleatoire
## Contexte

Audit de fidelite **#3164** (lot `research_*`). Finding **REINVENTION involontaire** sur `MyIA.AI.Notebooks/QuantConnect/research/research_rl_ppo.ipynb` (auteur : myia-po-2024, cycle anterieur #1454/#1461).

## Bug (confirme par lecture directe cell-6, G.1)

Dans `PPOAgent.update()` (cell-6), le bloc finite-difference de l'actor est **incomplet** :

```python
total_loss = 0            # ligne 318 : init, JAMAIS reassigne dans la boucle actor
eps = 0.001
params = self.actor.get_all_params()
grad = np.zeros_like(params)
for i in range(0, len(params), ...):
    ...
    loss_plus = ...       # loss aux params perturbes (correct)
    grad[i] = (loss_plus - total_loss) / eps if total_loss != 0 else 0   # ligne 339
if np.linalg.norm(grad) > 0:        # toujours False (grad == 0 partout)
    self.actor.set_all_params(params + self.actor.lr * grad)            # JAMAIS appele
```

Le guard `if total_loss != 0 else 0` est le smoking gun : check defensif pour une valeur **jamais calculee**. `total_loss` devait etre la baseline loss `L(theta)` aux params actuels (point de reference de la finite-difference), mais elle reste a 0. Resultat : `grad` est identiquement nul, l'actor n'est **jamais** mis a jour pendant tout l'entrainement (200 episodes x 4 seeds). Le critic, lui, s'update (boucle TD separee, ~ligne 358).

## Preuve
- `grep -n "total_loss" research_rl_ppo.ipynb` : 1 init (ligne 318), 0 reassignation dans la boucle actor, 1 return (ligne 361, retourne 0).
- Lecture cell-6 : confirme l'absence de calcul de baseline loss.

## Consequence fidelite (#3164)

Le "PPO" entraine = **policy aleatoire figee a l'init**. Le verdict pedagogique `NO BEATS` (Sharpe mean -0.338 vs B&H 1.417, 0/4 seeds positifs, sigma edge -0.31) est **correct par accident** mais pour la **mauvaise raison** : il mesure une policy non-entrainee, pas PPO. La conclusion "PPO ne bat pas B&H sur SPY" est donc **invalide** — non-fidele au portage de Schulman et al. 2017, qui requiert une ascent effective sur L^CLIP.

Le notebook est **partiellement honnete** : il note "approximation grossiere du gradient" (cell-15) et "prototype pedagogique" (cell-16), suggere PyTorch pour RL-03+. Mais "approximation grossiere" sous-entend "apprend mal" — alors qu'en realite l'actor **n'apprend jamais**. La distinction est pedagogiquement cruciale.

## Fix propose (chirurgical, ~10 lignes, cell-6)

Calculer la baseline `total_loss` (loss aux params actuels) **avant** la boucle finite-diff, et retirer le guard cassant :

```python
# Baseline loss aux params actuels (point de reference finite-difference)
total_loss = 0
for s, a, old_p, adv_val in zip(states, actions, old_probs, adv):
    logits = self.actor.forward(s)
    probs = softmax(logits)
    new_p = max(probs[a], 1e-8)
    ratio = new_p / max(old_p, 1e-8)
    clipped = min(max(ratio, 1 - self.clip_eps), 1 + self.clip_eps)
    total_loss += min(ratio * adv_val, clipped * adv_val)
eps = 0.001
params = self.actor.get_all_params()
grad = np.zeros_like(params)
for i in range(0, len(params), max(1, len(params) // 50)):
    ...
    grad[i] = (loss_plus - total_loss) / eps   # guard retire
```

L'actor apprendra rellement (finite-diff bruite). Overhead negligeable (1 forward-loop/update vs 50 existantes). Re-execution 4 seeds x 200 episodes (numpy, ~minutes). Verdict honnete : probablement toujours NO BEATS (finite-diff PPO sur signal financier faible), mais pour la **bonne** raison cette fois.

## Conformite
- Anti-regression : le fix **restaure l'intention** (finite-diff policy gradient), ne remplace pas par sorry/stub. Conforme.
- C.2/H.3 : re-execution complete apres modif cellule code.

See #3164

### PR DESCRIPTION
**Pull request by jsboige**
## Contexte

Audit de fidélité **#3164** (lot `research_*.ipynb`). Finding **REINVENTION involontaire** sur `research_rl_ppo.ipynb` (auteur initial : myia-po-2024, cycle #1454/#1461). Issue fille **#3326**.

## Bug (2 causes complementaires, cell-6)

Le PPO numpy n'entrainait **jamais** son actor — la policy restait figee a l'init aleatoire pour les 200 episodes x 4 seeds.

**Bug 1 (cause racine) — `SmallMLP.set_all_params`** : rebind `self.params[i]` vers de nouveaux arrays mais ne resync **pas** les attributs `self.W1/b1/W2/b2` que `forward()` lit reellement. Les perturbations finite-difference etaient donc invisibles a `forward()` -> `loss_plus == total_loss` -> `grad == 0`. Preuve directe : `flat[0] += 999` puis `set_all_params(flat)` donne `self.params[0][0,0]=999.013` mais `self.W1[0,0]=0.0126` (inchange), `params[0] is W1` = False.

**Bug 2 — baseline `total_loss`** : initialisee a 0, jamais calculee dans la boucle actor ; un guard defensif `if total_loss != 0 else 0` forcait alors `grad = 0` dans tous les cas.

Ces 2 bugs se masquaient mutuellement : corriger l'un sans l'autre laissait l'actor fige.

## Fix (chirurgical, cell-6, restore l'intention — pas de sorry/stub)

1. `set_all_params` resync les attributs : `self.W1, self.b1, self.W2, self.b2 = self.params` apres la boucle.
2. `total_loss` (baseline L(theta)) calculee avant la boucle finite-diff ; guard cassant retire.

**Sanity-check** : actor W1 max|delta| = **2.47e-04** apres 1 update (vs 0.0 avant) -> l'actor fait maintenant du gradient ascent.

## Validation (H.1 — execution reelle)

Papermill end-to-end, kernel python3, 209s, 0 erreur, exec_count 1-9 coherent, outputs presents :

| Metrique | Avant fix (policy figee) | Apres fix (PPO apprend) |
|---|---|---|
| Seed 0 | Sharpe 1.066 | Sharpe 1.055 |
| Seed 1 | -1.483 | -1.525 |
| Seed 7 | -0.156 | -0.216 |
| Seed 42 | -0.779 | -0.832 |
| **Mean** | **-0.338** | **-0.380** |
| **Verdict** | NO BEATS | **NO BEATS** |

**Transparence (G.1/G.2)** : les nombres sont quasi-identiques avant/apres. Le fix restaure la **correction** (l'actor fait du gradient ascent), pas le verdict. Avant : verdict NO BEATS correct **par accident** (policy aleatoire figee). Apres : verdict NO BEATS **legitime** (un PPO qui apprend reellement mais dont l'approximation finite-diff numpy est trop grossiere pour ameliorer sur B&H — la caveat honnete existante du notebook, cell-15/16). C.1 : aucun `raise NotImplementedError`/erreur volontaire.

## Conformite

- **Anti-regression** : restore l'intention (finite-diff policy gradient), pas un stub. Conforme.
- **C.2/H.3** : re-execution complete apres modif cellule code, outputs coherents committes.
- **Catalogue** : byte-identique a `main` (non regenere sur la branche).
- **Scope** : 1 notebook, 1 cellule (cell-6). Atomique.

Closes #3326
See #3164

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
[Hermes] research_rl_ppo.ipynb fix (#3326 frozen actor)

Security: ✅ No tokens/secrets
Root cause: params rebinding invisible to W1/b1/W2/b2 → grad=0 → actor frozen
Fix: self.W1, self.b1, self.W2, self.b2 = self.params resyncs after set_all_params + baseline loss now computed (was 0)
Notebook: re-executed, outputs consistent (state range shift expected)

Minor: BATCH_MODE = "true" (string) — if downstream checks truthiness, always True. Consider BATCH_MODE = True or confirm all consumers handle strings.

No blocking issues.

**Comment by clusterManager-Myia** [comment]
**[NanoClaw]** — LGTM

- PPO research notebook: repairs frozen PPO actor issue (#3326)
- +351/-67 — substantial repair with real execution outputs (counts 4–8)
- Correct fix pattern for gradient flow in PPO actor network

No concerns.
